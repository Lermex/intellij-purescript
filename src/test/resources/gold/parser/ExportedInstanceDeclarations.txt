Purescript File
  PsiElement(whitespace)('-- Tests that instances for non-exported classes / types do not appear in the\n-- result of `exportedDeclarations`.\n')
  PSASTWrapperElement(Program)
    PSASTWrapperElement(Module)
      PsiElement(module)('module')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(ModuleName)
        PSASTWrapperElement(Qualified)
          PsiElement(proper name)('Main')
        PsiElement(whitespace)('\n  ')
      PsiElement(()('(')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PsiElement(proper name)('Const')
        PsiElement(()('(')
        PsiElement(..)('..')
        PsiElement())(')')
        PsiElement(whitespace)('\n  ')
      PsiElement(,)(',')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PsiElement(class)('class')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)('\n  ')
      PsiElement(,)(',')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PSASTWrapperElement(ValueRef)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)('\n  ')
      PsiElement(,)(',')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PSASTWrapperElement(ValueRef)
          PsiElement(identifier)('main')
          PsiElement(whitespace)('\n  ')
      PsiElement())(')')
      PsiElement(whitespace)(' ')
      PsiElement(where)('where')
      PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(ImportDeclaration)
        PsiElement(import)('import')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ImportModuleName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Prelude')
          PsiElement(whitespace)('\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(ImportDeclaration)
        PsiElement(import)('import')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ImportModuleName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Control')
            PsiElement(.)('.')
            PsiElement(proper name)('Monad')
            PsiElement(.)('.')
            PsiElement(proper name)('Eff')
            PsiElement(.)('.')
            PsiElement(proper name)('Console')
          PsiElement(whitespace)(' ')
        PsiElement(()('(')
        PSASTWrapperElement(PositionedDeclarationRef)
          PSASTWrapperElement(ValueRef)
            PsiElement(identifier)('log')
        PsiElement())(')')
        PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(DataDeclaration)
        PsiElement(data)('data')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PsiElement(proper name)('Const')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeArgs)
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('b')
            PsiElement(whitespace)(' ')
        PsiElement(=)('=')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PsiElement(proper name)('Const')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('a')
            PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(TypeClassDeclaration)
        PsiElement(class)('class')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(Implies)
          <empty list>
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(TypeDeclaration)
          PSASTWrapperElement(TypeAnnotationName)
            PsiElement(identifier)('foo')
            PsiElement(whitespace)(' ')
          PsiElement(::)('::')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(GenericIdentifier)
                PsiElement(identifier)('a')
                PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(DataDeclaration)
        PsiElement(data)('data')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PsiElement(proper name)('NonexportedType')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeArgs)
          <empty list>
        PsiElement(=)('=')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PsiElement(proper name)('NonexportedType')
          PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(TypeClassDeclaration)
        PsiElement(class)('class')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('NonexportedClass')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(Implies)
          <empty list>
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(TypeDeclaration)
          PSASTWrapperElement(TypeAnnotationName)
            PsiElement(identifier)('notExported')
            PsiElement(whitespace)(' ')
          PsiElement(::)('::')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(GenericIdentifier)
                PsiElement(identifier)('a')
                PsiElement(whitespace)('\n\n-- There are three places that a nonexported type or type class can occur,\n-- leading an instance to count as non-exported:\n--  * Constraints\n--  * The type class itself\n--  * The instance types\n\n-- Case 1: constraints\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(GenericIdentifier)
          PsiElement(identifier)('nonExportedFoo')
          PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PsiElement(()('(')
        PSASTWrapperElement(TypeConstructor)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('NonexportedClass')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('a')
        PsiElement())(')')
        PsiElement(whitespace)(' ')
        PsiElement(=>)('=>')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(GenericIdentifier)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(NamedBinder)
            <empty list>
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Var)
                PSASTWrapperElement(Qualified)
                  PsiElement(identifier)('notExported')
                  PsiElement(whitespace)('\n\n-- Another instance of case 1:\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(GenericIdentifier)
          PsiElement(identifier)('nonExportedFoo2')
          PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PsiElement(()('(')
        PSASTWrapperElement(TypeConstructor)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeConstructor)
            PSASTWrapperElement(Qualified)
              PsiElement(proper name)('NonexportedType')
        PsiElement())(')')
        PsiElement(whitespace)(' ')
        PsiElement(=>)('=>')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PsiElement(()('(')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(GenericIdentifier)
                PsiElement(identifier)('a')
                PsiElement(whitespace)(' ')
            PsiElement(->)('->')
            PsiElement(whitespace)(' ')
            PSASTWrapperElement(Type)
              PSASTWrapperElement(TypeAtom)
                PSASTWrapperElement(GenericIdentifier)
                  PsiElement(identifier)('a')
          PsiElement())(')')
          PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(NamedBinder)
            <empty list>
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Var)
                PSASTWrapperElement(Qualified)
                  PsiElement(identifier)('id')
                  PsiElement(whitespace)('\n\n-- Case 2: type class\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(GenericIdentifier)
          PsiElement(identifier)('nonExportedNonexportedType')
          PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('NonexportedClass')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PsiElement(()('(')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Int')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(GenericIdentifier)
                PsiElement(identifier)('a')
          PsiElement())(')')
          PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('notExported')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(NamedBinder)
            <empty list>
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Constructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
              PSASTWrapperElement(NumericLiteral)
                PsiElement(natural)('0')
                PsiElement(whitespace)('\n\n-- Case 3: instance types\n')
      PSASTWrapperElement(NamedBinder)
        <empty list>
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(GenericIdentifier)
          PsiElement(identifier)('constFoo')
          PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeConstructor)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PsiElement(()('(')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('NonexportedType')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(GenericIdentifier)
                PsiElement(identifier)('b')
          PsiElement())(')')
          PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(NamedBinder)
            <empty list>
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Constructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
              PSASTWrapperElement(Constructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('NonexportedType')
                  PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(ValueDeclaration)
        PsiElement(identifier)('main')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(NamedBinder)
          <empty list>
        PsiElement(=)('=')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(Value)
          PSASTWrapperElement(PrefixValue)
            PSASTWrapperElement(Var)
              PSASTWrapperElement(Qualified)
                PsiElement(identifier)('log')
                PsiElement(whitespace)(' ')
            PSASTWrapperElement(StringLiteral)
              PsiElement(string)('"Done"')