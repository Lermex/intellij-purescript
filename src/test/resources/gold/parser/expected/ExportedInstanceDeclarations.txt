Purescript File
  PsiElement(whitespace)('-- Tests that instances for non-exported classes / types do not appear in the\n-- result of `exportedDeclarations`.\n')
  PSASTWrapperElement(Program)
    PSASTWrapperElement(Module)
      PsiElement(module)('module')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(Qualified)
        PSASTWrapperElement(ModuleName)
          PsiElement(proper name)('Main')
      PsiElement(whitespace)('\n  ')
      PsiElement(()('(')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PsiElement(proper name)('Const')
        PsiElement(()('(')
        PsiElement(..)('..')
        PsiElement())(')')
        PsiElement(whitespace)('\n  ')
      PsiElement(,)(',')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PsiElement(class)('class')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)('\n  ')
      PsiElement(,)(',')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PSASTWrapperElement(ValueRef)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)('\n  ')
      PsiElement(,)(',')
      PsiElement(whitespace)(' ')
      PSASTWrapperElement(PositionedDeclarationRef)
        PSASTWrapperElement(ValueRef)
          PsiElement(identifier)('main')
          PsiElement(whitespace)('\n  ')
      PsiElement())(')')
      PsiElement(whitespace)(' ')
      PsiElement(where)('where')
      PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(ImportDeclaration)
        PsiElement(import)('import')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(Qualified)
          PSASTWrapperElement(ModuleName)
            PsiElement(proper name)('Prelude')
        PsiElement(whitespace)('\n')
      PSASTWrapperElement(ImportDeclaration)
        PsiElement(import)('import')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(Qualified)
          PSASTWrapperElement(ModuleName)
            PsiElement(proper name)('Control')
            PsiElement(.)('.')
          PSASTWrapperElement(ModuleName)
            PsiElement(proper name)('Monad')
            PsiElement(.)('.')
          PSASTWrapperElement(ModuleName)
            PsiElement(proper name)('Eff')
            PsiElement(.)('.')
          PSASTWrapperElement(ModuleName)
            PsiElement(proper name)('Console')
        PsiElement(whitespace)(' ')
        PsiElement(()('(')
        PSASTWrapperElement(PositionedDeclarationRef)
          PSASTWrapperElement(ValueRef)
            PsiElement(identifier)('log')
        PsiElement())(')')
        PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(DataDeclaration)
        PsiElement(data)('data')
        PsiElement(whitespace)(' ')
        PsiElement(proper name)('Const')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeArgs)
          PsiElement(identifier)('a')
          PsiElement(whitespace)(' ')
          PsiElement(identifier)('b')
          PsiElement(whitespace)(' ')
        PsiElement(=)('=')
        PsiElement(whitespace)(' ')
        PsiElement(proper name)('Const')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeVar)
            PsiElement(identifier)('a')
            PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(TypeClassDeclaration)
        PsiElement(class)('class')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(Qualified)
          PsiElement(proper name)('Foo')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeVar)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(Implies)
          <empty list>
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(TypeDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PsiElement(::)('::')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeVar)
                PsiElement(identifier)('a')
                PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(DataDeclaration)
        PsiElement(data)('data')
        PsiElement(whitespace)(' ')
        PsiElement(proper name)('NonexportedType')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeArgs)
          <empty list>
        PsiElement(=)('=')
        PsiElement(whitespace)(' ')
        PsiElement(proper name)('NonexportedType')
        PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(TypeClassDeclaration)
        PsiElement(class)('class')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(Qualified)
          PsiElement(proper name)('NonexportedClass')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeVar)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(Implies)
          <empty list>
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(TypeDeclaration)
          PsiElement(identifier)('notExported')
          PsiElement(whitespace)(' ')
          PsiElement(::)('::')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeVar)
                PsiElement(identifier)('a')
                PsiElement(whitespace)('\n\n-- There are three places that a nonexported type or type class can occur,\n-- leading an instance to count as non-exported:\n--  * Constraints\n--  * The type class itself\n--  * The instance types\n\n-- Case 1: constraints\n')
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PsiElement(identifier)('nonExportedFoo')
        PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PsiElement(()('(')
        PSASTWrapperElement(Qualified)
          PsiElement(proper name)('NonexportedClass')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeVar)
            PsiElement(identifier)('a')
        PsiElement())(')')
        PsiElement(whitespace)(' ')
        PsiElement(=>)('=>')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeVar)
            PsiElement(identifier)('a')
            PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Var)
                PSASTWrapperElement(Qualified)
                  PsiElement(identifier)('notExported')
                  PsiElement(whitespace)('\n\n-- Another instance of case 1:\n')
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PsiElement(identifier)('nonExportedFoo2')
        PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PsiElement(()('(')
        PSASTWrapperElement(Qualified)
          PsiElement(proper name)('Foo')
          PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PSASTWrapperElement(TypeConstructor)
            PSASTWrapperElement(Qualified)
              PsiElement(proper name)('NonexportedType')
        PsiElement())(')')
        PsiElement(whitespace)(' ')
        PsiElement(=>)('=>')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PsiElement(()('(')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeVar)
                PsiElement(identifier)('a')
                PsiElement(whitespace)(' ')
            PsiElement(->)('->')
            PsiElement(whitespace)(' ')
            PSASTWrapperElement(Type)
              PSASTWrapperElement(TypeAtom)
                PSASTWrapperElement(TypeVar)
                  PsiElement(identifier)('a')
          PsiElement())(')')
          PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Var)
                PSASTWrapperElement(Qualified)
                  PsiElement(identifier)('id')
                  PsiElement(whitespace)('\n\n-- Case 2: type class\n')
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PsiElement(identifier)('nonExportedNonexportedType')
        PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('NonexportedClass')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PsiElement(()('(')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Int')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeVar)
                PsiElement(identifier)('a')
          PsiElement())(')')
          PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('notExported')
          PsiElement(whitespace)(' ')
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Constructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
              PSASTWrapperElement(NumericLiteral)
                PsiElement(natural)('0')
                PsiElement(whitespace)('\n\n-- Case 3: instance types\n')
      PSASTWrapperElement(TypeInstanceDeclaration)
        PsiElement(instance)('instance')
        PsiElement(whitespace)(' ')
        PsiElement(identifier)('constFoo')
        PsiElement(whitespace)(' ')
        PsiElement(::)('::')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(ClassName)
          PSASTWrapperElement(Qualified)
            PsiElement(proper name)('Foo')
            PsiElement(whitespace)(' ')
        PSASTWrapperElement(TypeAtom)
          PsiElement(()('(')
          PSASTWrapperElement(Type)
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeConstructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('NonexportedType')
                  PsiElement(whitespace)(' ')
            PSASTWrapperElement(TypeAtom)
              PSASTWrapperElement(TypeVar)
                PsiElement(identifier)('b')
          PsiElement())(')')
          PsiElement(whitespace)(' ')
        PsiElement(where)('where')
        PsiElement(whitespace)('\n  ')
        PSASTWrapperElement(ValueDeclaration)
          PsiElement(identifier)('foo')
          PsiElement(whitespace)(' ')
          PsiElement(=)('=')
          PsiElement(whitespace)(' ')
          PSASTWrapperElement(Value)
            PSASTWrapperElement(PrefixValue)
              PSASTWrapperElement(Constructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('Const')
                  PsiElement(whitespace)(' ')
              PSASTWrapperElement(Constructor)
                PSASTWrapperElement(Qualified)
                  PsiElement(proper name)('NonexportedType')
                  PsiElement(whitespace)('\n\n')
      PSASTWrapperElement(ValueDeclaration)
        PsiElement(identifier)('main')
        PsiElement(whitespace)(' ')
        PsiElement(=)('=')
        PsiElement(whitespace)(' ')
        PSASTWrapperElement(Value)
          PSASTWrapperElement(PrefixValue)
            PSASTWrapperElement(Var)
              PSASTWrapperElement(Qualified)
                PsiElement(identifier)('log')
                PsiElement(whitespace)(' ')
            PSASTWrapperElement(StringLiteral)
              PsiElement(string)('"Done"')